gradio/route_utils.py

```diff
@@ -264,3 +270,189 @@ def strip_url(orig_url: str) -> str:
     stripped_url = parsed_url.copy_with(query=None)
     stripped_url = str(stripped_url)
     return stripped_url.rstrip("/")
+
+
+def _user_safe_decode(src: bytes, codec: str) -> str:
+    try:
+        return src.decode(codec)
+    except (UnicodeDecodeError, LookupError):
+        return src.decode("latin-1")
+
+
+class GradioUploadFile(UploadFile):
+    """UploadFile with a sha attribute."""
+
+    def __init__(
+        self,
+        file: BinaryIO,
+        *,
+        size: int | None = None,
+        filename: str | None = None,
+        headers: Headers | None = None,
+    ) -> None:
+        super().__init__(file, size=size, filename=filename, headers=headers)
+        self.sha = hashlib.sha1()
+
+
+class GradioMultiPartParser:
```

Comment:
This is vendored from starlette. It's also possible to do this via subclassing MultipartParser from starlette but that leaves us vulnerable to any changes starlette makes to that class.